include/morphio/collection.h

```diff
@@ -61,10 +85,94 @@ class Collection
     std::shared_ptr<CollectionImpl> _collection;
 };
 
-extern template mut::Morphology Collection::load<mut::Morphology>(const std::string& morph_name,
-                                                                  unsigned int options) const;
+class LoadUnorderedImpl;
+class LoadUnorderedIteratorImpl;
+
+/**
+ * An iterable of loop index and morphologies.
+ *
+ * When reading from containers, the order in which morphologies are read can
+ * have a large impact on the overall time to load those morphologies.
+ *
+ * This iterator provides means of reordering loops to optimize the access
+ * pattern. Loops such as the following
+ *
+ *     for(size_t k = 0; k < morphology_names.size; ++k) {
+ *       auto morph = collection.load<M>(morphology_names[k]);
+ *       f(k, morph);
+ *     }
+ *
+ * can be replaced with
+ *
+ *     for(auto [k, morph] : collection.load_unordered<M>(morphology_names)) {
+ *       assert(collection.load<M>(morphology_names[k]) == morph);
+ *       f(k, morph);
+ *     }
+ *
+ * The order in which the morphologies are returned in unspecified, but the
+ * loop index `k` can be used to retrieve the correct state corresponding to
+ * iteration `k` of the original loop.
+ *
+ * Note, that it is safe for an `LoadUnordered` object to outlive its
+ * `collection`. Internally a shallow copy of the original `collection` is
+ * stored inside of and kept alive for the life time of the `LoadUnordered`
+ * object.
+ */
+template <class M>
+class LoadUnordered
+{
+  protected:
+    class Iterator
+    {
+      public:
+        Iterator(std::shared_ptr<LoadUnorderedIteratorImpl> it);
+
+        template <class U = M>
+        typename enable_if_immutable<U, std::pair<size_t, M>>::type operator*() const;
+
+        template <class U = M>
+        typename enable_if_mutable<U, std::pair<size_t, M>>::type operator*() const;
+
+        void operator++() const;
```

Comment:
prob. not important, but I tend to like the symmetry of having `operator++(int)`

Reply:
We've had to redo the `LoadUnordered<M>::Iterator`, it had multiple bugs related to unintended shallow copying. Tests have been added to check for this type of issue.